web/src/app/Notifier.js

```diff
@@ -1,39 +1,34 @@
-import { openUrl, playSound, topicDisplayName, topicShortUrl, urlB64ToUint8Array } from "./utils";
-import { formatMessage, formatTitleWithDefault } from "./notificationUtils";
+import { playSound, topicDisplayName, topicShortUrl, urlB64ToUint8Array } from "./utils";
+import { getNotificationParams } from "./notificationUtils";
 import prefs from "./Prefs";
-import logo from "../img/ntfy.png";
+import routes from "../components/routes";
 
 /**
  * The notifier is responsible for displaying desktop notifications. Note that not all modern browsers
  * support this; most importantly, all iOS browsers do not support window.Notification.
  */
 class Notifier {
-  async notify(subscription, notification, onClickFallback) {
+  async notify(subscription, notification) {
     if (!this.supported()) {
       return;
     }
 
+    await this.playSound();
+
     const shortUrl = topicShortUrl(subscription.baseUrl, subscription.topic);
-    const displayName = topicDisplayName(subscription);
-    const message = formatMessage(notification);
-    const title = formatTitleWithDefault(notification, displayName);
-    const image = notification.attachment?.name.match(/\.(png|jpe?g|gif|webp)$/i) ? notification.attachment.url : undefined;
-
-    // Show notification
-    console.log(`[Notifier, ${shortUrl}] Displaying notification ${notification.id}: ${message}`);
-    // Please update sw.js if formatting changes
-    const n = new Notification(title, {
-      body: message,
-      tag: subscription.id,
-      icon: image ?? logo,
-      image,
-      timestamp: message.time * 1_000,
-    });
-    if (notification.click) {
-      n.onclick = () => openUrl(notification.click);
-    } else {
-      n.onclick = () => onClickFallback(subscription);
-    }
+    const defaultTitle = topicDisplayName(subscription);
+
+    console.log(`[Notifier, ${shortUrl}] Displaying notification ${notification.id}`);
+
+    const registration = await this.serviceWorkerRegistration();
+    await registration.showNotification(
+      ...getNotificationParams({
+        subscriptionId: subscription.id,
+        message: notification,
+        defaultTitle,
+        topicRoute: new URL(routes.forSubscription(subscription), window.location.origin).toString(),
+      })
+    );
```

Comment:
If I understand correctly, this now implies that the service worker is correctly registered to display notifications in the normal web app. What is this change about? Just to harmonize things? What is the risk here?

I sadly cannot test half the features, because I'm on Linux and there are no action buttons and such.

Reply:
yes - at first I left it alone but I've been testing it over the last few weeks and the sw setup is stable.

- The main advantage is making them work the same and not diverge in behaviour. 
- The downside is not working where service workers do not work, but that's not really a risk: https://caniuse.com/serviceworkers - the one thing that I've seen is that Firefox Private Mode doesn't support them, but ntfy currently doesn't work there anyway due to IndexedDB.

P.S. you can test actions on Chrome, on Linux too: https://developer.mozilla.org/en-US/docs/Web/API/Notification/actions#browser_compatibility

Reply:
yes - at first I left it alone but I've been testing it over the last few weeks and the sw setup is stable.

- The main advantage is making them work the same and not diverge in behaviour.
- The downside is not working where service workers do not work, but that's not really a risk: <https://caniuse.com/serviceworkers> - the one thing that I've seen is that Firefox Private Mode doesn't support them, but ntfy currently doesn't work there anyway due to IndexedDB.

P.S. you can test actions on Chrome, on Linux too: <https://developer.mozilla.org/en-US/docs/Web/API/Notification/actions#browser_compatibility>

Reply:
Thanks for the explanation! I think it's fine. Let's see if we break anything :grimacing: